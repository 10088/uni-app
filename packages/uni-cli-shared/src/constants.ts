export const EXTNAME_JS = ['.js', '.ts']
export const EXTNAME_VUE = ['.vue', '.nvue', '.fvue']
