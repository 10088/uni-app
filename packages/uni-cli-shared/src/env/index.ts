export { initDefine } from './define'
export { initProvide } from './provide'
