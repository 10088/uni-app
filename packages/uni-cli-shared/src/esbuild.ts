import { TransformOptions } from 'esbuild'

export function transformWithEsbuild(code: string, options: TransformOptions) {
  return import('esbuild').then((esbuild) => {
    return esbuild.transform(code, options)
  })
}
