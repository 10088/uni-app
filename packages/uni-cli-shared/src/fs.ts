import fs from 'fs'
import path from 'path'
export function emptyDir(dir: string, skip?: string[]): void {
  for (const file of fs.readdirSync(dir)) {
    if (skip?.includes(file)) {
      continue
    }
    const abs = path.resolve(dir, file)
    // baseline is Node 12 so can't use rmSync :(
    if (fs.lstatSync(abs).isDirectory()) {
      emptyDir(abs)
      fs.rmdirSync(abs)
    } else {
      fs.unlinkSync(abs)
    }
  }
}
