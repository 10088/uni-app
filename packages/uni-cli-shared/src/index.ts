export * from './json'
export * from './easycom'
export * from './vueCompilerOptions'
export * from './transforms/transformBuiltInComponent'
export * from './transforms/transformBuiltInEasycom'
export * from './transforms/transformEasycom'
export { isAutoImported } from './transforms/autoImport'
