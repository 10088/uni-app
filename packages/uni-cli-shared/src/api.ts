import { BASE_COMPONENTS_STYLE_PATH, H5_API_STYLE_PATH } from './constants'

export const API_STYLES = {
  showModal: [`${H5_API_STYLE_PATH}/modal.css`],
  showToast: [`${H5_API_STYLE_PATH}/toast.css`],
  showActionSheet: [`${H5_API_STYLE_PATH}/action-sheet.css`],
  previewImage: [
    `${BASE_COMPONENTS_STYLE_PATH}/swiper.css`,
    `${BASE_COMPONENTS_STYLE_PATH}/swiper-item.css`,
    `${BASE_COMPONENTS_STYLE_PATH}/movable-area.css`,
    `${BASE_COMPONENTS_STYLE_PATH}/movable-view.css`,
  ],
  openLocation: [`${H5_API_STYLE_PATH}/location-view.css`],
}
