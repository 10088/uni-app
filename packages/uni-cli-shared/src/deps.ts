import { BASE_COMPONENTS_STYLE_PATH, H5_API_STYLE_PATH } from './constants'

export const API_DEPS_CSS = {
  showModal: [`${H5_API_STYLE_PATH}modal.css`],
  showToast: [`${H5_API_STYLE_PATH}toast.css`],
  showActionSheet: [`${H5_API_STYLE_PATH}action-sheet.css`],
  previewImage: [
    `${BASE_COMPONENTS_STYLE_PATH}swiper.css`,
    `${BASE_COMPONENTS_STYLE_PATH}swiper-item.css`,
    `${BASE_COMPONENTS_STYLE_PATH}movable-area.css`,
    `${BASE_COMPONENTS_STYLE_PATH}movable-view.css`,
  ],
  openLocation: [`${H5_API_STYLE_PATH}location-view.css`],
}

const RESIZE_SENSOR_CSS = BASE_COMPONENTS_STYLE_PATH + 'resize-sensor.css'

export const COMPONENT_DEPS_CSS = {
  canvas: [RESIZE_SENSOR_CSS],
  image: [RESIZE_SENSOR_CSS],
  'movable-area': [RESIZE_SENSOR_CSS],
  'picker-view': [RESIZE_SENSOR_CSS],
  'picker-view-column': [RESIZE_SENSOR_CSS],
  'rich-text': [RESIZE_SENSOR_CSS],
  textarea: [RESIZE_SENSOR_CSS],
  'web-view': [RESIZE_SENSOR_CSS],
  picker: [
    RESIZE_SENSOR_CSS,
    `${BASE_COMPONENTS_STYLE_PATH}picker-view.css`,
    `${BASE_COMPONENTS_STYLE_PATH}picker-view-column.css`,
  ],
}
