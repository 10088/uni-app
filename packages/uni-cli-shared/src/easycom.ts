import fs from 'fs'
import path from 'path'

interface EasycomOption {
  dirs?: string[]
  rootDir?: string
  custom?: EasycomCustom
}
interface EasycomMatcher {
  pattern: RegExp
  replacement: string
}
interface EasycomCustom {
  [key: string]: string
}

const easycoms: EasycomMatcher[] = []

const easycomsCache = new Map<string, string>()
const easycomsInvalidCache = new Set<string>()

function clearEasycoms() {
  easycoms.length = 0
  easycomsCache.clear()
  easycomsInvalidCache.clear()
}

export function initEasycoms({ dirs, rootDir, custom }: EasycomOption) {
  clearEasycoms()
  const easycomsObj = Object.create(null)
  if (dirs && rootDir) {
    Object.assign(easycomsObj, initAutoScanEasycoms(dirs, rootDir))
  }
  if (custom) {
    Object.assign(easycomsObj, custom)
  }
  Object.keys(easycomsObj).forEach(name => {
    easycoms.push({
      pattern: new RegExp(name),
      replacement: easycomsObj[name]
    })
  })
  return easycoms
}

export function matchEasycom(tag: string) {
  let source = easycomsCache.get(tag)
  if (source) {
    return source
  }
  if (easycomsInvalidCache.has(tag)) {
    return false
  }
  const matcher = easycoms.find(matcher => matcher.pattern.test(tag))
  if (!matcher) {
    easycomsInvalidCache.add(tag)
    return false
  }
  source = tag.replace(matcher.pattern, matcher.replacement)
  easycomsCache.set(tag, source)
  return source
}

const isDir = (path: string) => fs.lstatSync(path).isDirectory()

function initAutoScanEasycom(
  dir: string,
  rootDir: string
): Record<string, string> {
  const easycoms = Object.create(null)
  if (!fs.existsSync(dir)) {
    return easycoms
  }
  fs.readdirSync(dir).forEach(name => {
    const folder = path.resolve(dir, name)
    if (!isDir(folder)) {
      return
    }
    const importDir = path.relative(rootDir, folder)
    const files = fs.readdirSync(folder)
    if (files.find(file => path.parse(file).name === name)) {
      easycoms[`^${name}$`] = `@/${importDir}/${name}`
    }
  })
  return easycoms
}

function initAutoScanEasycoms(dirs: string[], rootDir: string) {
  return dirs.reduce<Record<string, string>>(
    (easycoms: Record<string, string>, dir: string) => {
      const curEasycoms = initAutoScanEasycom(dir, rootDir)
      Object.keys(curEasycoms).forEach(name => {
        // Use the first component when name conflict
        if (!easycoms[name]) {
          easycoms[name] = curEasycoms[name]
        }
      })
      return easycoms
    },
    Object.create(null)
  )
}
