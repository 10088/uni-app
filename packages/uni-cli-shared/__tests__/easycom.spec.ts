import path from 'path'

import { parse, compileTemplate, compileScript } from '@vue/compiler-sfc'

import { initEasycom, matchEasycom, vueCompilerOptions } from '../src'

const rootDir = path.resolve(__dirname, 'example')
const dirs = [
  path.resolve(__dirname, 'example/components'),
  path.resolve(__dirname, 'example/uni_modules/plugin/components')
]

const template = `<template><test/><test1/><test1/><test2/><some-other-comp/><checkbox-group/><checkbox-group/></template>`
const sfcParseRes = parse(`${template}<script>export default {}</script>`)
const sfcParseResWithSetup = parse(
  `${template}<script setup>import SomeOtherComp from './some-other-comp';import Test from './test';</script>`
)
const id = 'test'
const sfcScriptCompileOptions = { id }
const sfcScriptBlock = compileScript(
  sfcParseRes.descriptor,
  sfcScriptCompileOptions
)
const sfcScriptBlockWithSetup = compileScript(
  sfcParseResWithSetup.descriptor,
  sfcScriptCompileOptions
)

describe('easycom', () => {
  test('initEasycom with dirs', () => {
    expect(initEasycom({ dirs, rootDir })).toEqual([
      {
        pattern: new RegExp('^test$'),
        replacement: '@/components/test/test.vue'
      },
      {
        pattern: new RegExp('^test1$'),
        replacement: '@/components/test1/test1.vue'
      },
      {
        pattern: new RegExp('^test2$'),
        replacement: '@/uni_modules/plugin/components/test2/test2.vue'
      }
    ])
    expect(matchEasycom('test')).toBe('@/components/test/test.vue')
    expect(matchEasycom('test1')).toBe('@/components/test1/test1.vue')
    expect(matchEasycom('test2')).toBe(
      '@/uni_modules/plugin/components/test2/test2.vue'
    )
  })
  test('initEasycom with custom', () => {
    expect(
      initEasycom({ custom: { '^uni-(.*)': '@/components/uni-$1.vue' } })
    ).toEqual([
      {
        pattern: new RegExp('^uni-(.*)'),
        replacement: '@/components/uni-$1.vue'
      }
    ])
    expect(matchEasycom('test')).toBe(false)
    expect(matchEasycom('uni-test1')).toBe('@/components/uni-test1.vue')
  })
  test('initEasycom with dirs and custom', () => {
    expect(
      initEasycom({
        dirs,
        rootDir,
        custom: { '^test$': '@/components/uni-test.vue' }
      })
    ).toEqual([
      {
        pattern: new RegExp('^test$'),
        replacement: '@/components/uni-test.vue'
      },
      {
        pattern: new RegExp('^test1$'),
        replacement: '@/components/test1/test1.vue'
      },
      {
        pattern: new RegExp('^test2$'),
        replacement: '@/uni_modules/plugin/components/test2/test2.vue'
      }
    ])
    expect(matchEasycom('test')).toBe('@/components/uni-test.vue')
  })
  test('render', () => {
    initEasycom({
      dirs,
      rootDir,
      custom: { '^test$': '@/components/uni-test.vue' }
    })
    const { code } = compileTemplate(
      Object.assign(sfcParseRes.descriptor, {
        id,
        compilerOptions: Object.assign(
          {
            bindingMetadata: sfcScriptBlock.bindings
          },
          vueCompilerOptions
        )
      })
    )
    expect(code).toMatch(`  // const _component_test = _resolveComponent("test")
  // const _component_test1 = _resolveComponent("test1")
  // const _component_test2 = _resolveComponent("test2")
  const _component_some_other_comp = _resolveComponent("some-other-comp")
  // const _component_v_uni_checkbox_group = _resolveComponent("v-uni-checkbox-group")`)
    expect(code)
      .toMatch(`import _component_test from '@/components/uni-test.vue'
import _component_test1 from '@/components/test1/test1.vue'
import _component_test2 from '@/uni_modules/plugin/components/test2/test2.vue'
import { CheckboxGroup as _component_v_uni_checkbox_group } from '@dcloudio/uni-h5/dist/uni-h5.esm.js'
import _style_v_uni_checkbox_group from '@dcloudio/uni-h5/style/checkbox-group.css'`)
    expect(code).toMatchSnapshot()
  })
  test('render with setup', () => {
    initEasycom({
      dirs,
      rootDir,
      custom: { '^test$': '@/components/uni-test.vue' }
    })
    const { code } = compileTemplate(
      Object.assign(sfcParseResWithSetup.descriptor, {
        id,
        compilerOptions: Object.assign(
          { bindingMetadata: sfcScriptBlockWithSetup.bindings },
          vueCompilerOptions
        )
      })
    )
    expect(code)
      .toMatch(`  // const _component_test1 = _resolveComponent("test1")
  // const _component_test2 = _resolveComponent("test2")
  // const _component_v_uni_checkbox_group = _resolveComponent("v-uni-checkbox-group")`)
    expect(code).toMatchSnapshot()
  })
})
