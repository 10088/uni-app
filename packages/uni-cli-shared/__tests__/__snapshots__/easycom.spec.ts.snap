// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`easycom render 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"
import _component_test from '@/components/uni-test.vue'
import _component_test1 from '@/components/test1/test1.vue'
import _component_test2 from '@/uni_modules/plugin/components/test2/test2.vue'
import { CheckboxGroup as _component_v_uni_checkbox_group } from '@dcloudio/uni-h5/dist/uni-h5.esm.js'
import _style_v_uni_checkbox_group from '@dcloudio/uni-h5/style/checkbox-group.css'


export function render(_ctx, _cache, $props, $setup, $data, $options) {
  // const _component_test = _resolveComponent(\\"test\\")
  // const _component_test1 = _resolveComponent(\\"test1\\")
  // const _component_test2 = _resolveComponent(\\"test2\\")
  const _component_some_other_comp = _resolveComponent(\\"some-other-comp\\")
  // const _component_v_uni_checkbox_group = _resolveComponent(\\"v-uni-checkbox-group\\")

  return (_openBlock(), _createBlock(\\"template\\", null, [
    _createVNode(_component_test),
    _createVNode(_component_test1),
    _createVNode(_component_test1),
    _createVNode(_component_test2),
    _createVNode(_component_some_other_comp),
    _createVNode(_component_v_uni_checkbox_group),
    _createVNode(_component_v_uni_checkbox_group)
  ]))
}"
`;

exports[`easycom render with setup 1`] = `
"import { createVNode as _createVNode, resolveComponent as _resolveComponent, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"
import _component_test1 from '@/components/test1/test1.vue'
import _component_test2 from '@/uni_modules/plugin/components/test2/test2.vue'
import { CheckboxGroup as _component_v_uni_checkbox_group } from '@dcloudio/uni-h5/dist/uni-h5.esm.js'
import _style_v_uni_checkbox_group from '@dcloudio/uni-h5/style/checkbox-group.css'


export function render(_ctx, _cache, $props, $setup, $data, $options) {
  // const _component_test1 = _resolveComponent(\\"test1\\")
  // const _component_test2 = _resolveComponent(\\"test2\\")
  // const _component_v_uni_checkbox_group = _resolveComponent(\\"v-uni-checkbox-group\\")

  return (_openBlock(), _createBlock(\\"template\\", null, [
    _createVNode($setup[\\"Test\\"]),
    _createVNode(_component_test1),
    _createVNode(_component_test1),
    _createVNode(_component_test2),
    _createVNode($setup[\\"SomeOtherComp\\"]),
    _createVNode(_component_v_uni_checkbox_group),
    _createVNode(_component_v_uni_checkbox_group)
  ]))
}"
`;
